### PR TITLE
[BEAM-2359] Fix watermark broadcasting to executors in Spark runner

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
R: @amitsela 
CC: @staslev @kobisalant

### CODE REVIEW

**Comment by aviemzur** [comment]
Run ValidatesRunner Spark

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

File: runners/spark/src/test/java/org/apache/beam/runners/spark/GlobalWatermarkHolderTest.java [hunk 1]
```diff
@@ -65,17 +65,17 @@ public void testLowHighWatermarksAdvance() {
             instant.plus(Duration.millis(5)),
             instant.plus(Duration.millis(5)),
             instant));
-    GlobalWatermarkHolder.advance(jsc);
```
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/GlobalWatermarkHolderTest.java]: the fact that `JavaSparkContext` is not needed here at all anymore makes me happy 😄

File: runners/spark/src/main/java/org/apache/beam/runners/spark/util/GlobalWatermarkHolder.java [hunk 2]
```diff
@@ -119,28 +179,34 @@ public static void advance(JavaSparkContext jsc) {
...
-        }
-        broadcast = jsc.broadcast(newBroadcast);
+      if (!newValues.isEmpty()) {
+        blockManager.removeBlock(WATERMARKS_BLOCK_ID, true);
+        blockManager.putSingle(
+            WATERMARKS_BLOCK_ID,
+            newValues,
+            StorageLevel.MEMORY_ONLY(),
+            true);
+        String breaK = "point";
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/util/GlobalWatermarkHolder.java]: ?

**Comment by amitsela** [comment]
Run Spark ValidatesRunner

File: runners/spark/src/main/java/org/apache/beam/runners/spark/util/GlobalWatermarkHolder.java (hunk 2 above)
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/util/GlobalWatermarkHolder.java]: Left by mistake, will remove it

**Comment by aviemzur** [comment]
Retest this please

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Retest this please

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
We forcibly tell it to fetch the remote value if the value we cached using a Guava cache has expired: https://github.com/apache/beam/pull/3343/files#diff-90f8c925037a33f132339f9ada1e2e38R258
```java
SparkEnv.get().blockManager().getRemote(WATERMARKS_BLOCK_ID);
```

**Comment by aviemzur** [comment]
I'll go ahead and add the `batchInterval` to SparkRuntimeContext as suggested

**Comment by baibaichen** [comment]
Interesting, BlockManager  is declared package private class: `private[spark] class BlockManager`. But, we can access its member,e.g. `putSingle`.

**Comment by staslev** [comment]
@baibaichen, Scala's package private stuff are equivalent to public when accessed from Java.

**Comment by baibaichen** [comment]
@staslev thanks, I verified it. Yes, Scala does fail but java works.

@aviemzur @amitsela I have 2 questions for such workaroud:
1.  Compatibility,  for example `BlockManager#getRemote` was renamed to BlockManager#getRemoteValues` which is private.
2. Recovery from restart. I didn't see any info about this in the jira.

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
@baibaichen 

1. Regarding compatibility there is an ongoing effort for supporting Spark 2 by @jbonofre that will use something akin to SpEL to support both APIs, this will have to be included there as well. 
2. Regarding recovery from restart this is not a part of the original implementation either, once we start from a checkpoint, it will take 1 microbatch for the watermarks to get updated.

**Comment by amitsela** [comment]
LGMT once tests pass

**Comment by aviemzur** [comment]
Retest this please

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Retest this please

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Retest this please

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Retest this please

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Pull request closed without merging**